### PR TITLE
GG-25778: Extended test coverage for Partition Awareness

### PR DESCRIPTION
**Pull request by isapego**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,6 @@
 .idea
+node_modules
+build
+npm-debug.log
+/logs
+package-lock.json
```
Comment (ptupitsyn) [inline .gitignore:6]: AFAIK, `package-lock.json` is supposed to be in source control: 
https://stackoverflow.com/questions/44206782/do-i-commit-the-package-lock-json-file-created-by-npm-5

File: spec/affinity_awareness/AffinityAwarenessConnection.spec.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        Promise.resolve().
+            then(async () => {
+                await TestingHelper.cleanUp();
+            }).
+            then(done).
+            catch(_error => done());
+    }, TestingHelper.TIMEOUT);
+
+    it('client with affinity awareness and bad servers', (done) => {
+        Promise.resolve().
```
Comment (ptupitsyn) [inline spec/affinity_awareness/AffinityAwarenessConnection.spec.js:50]: Not sure I understand the mix of explicit Promise usage and async/await in these tests.
Normally we don't need explicit Promises - async/await is enough, and results in cleaner code: https://jasmine.github.io/tutorials/async

File: spec/affinity_awareness/AffinityAwarenessConnection.spec.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                    expect(badEndpoints).not.toContain(ep);
+
+                const client = TestingHelper.makeClient();
+                const cfg = new IgniteClientConfiguration(...badEndpoints).setConnectionOptions(false, null, true);
+
+                try {
+                    await client.connect(cfg);
+                }
+                catch (error) {
+                    expect(error.stack).toContain('Connection failed');
```
Comment (ptupitsyn) [inline spec/affinity_awareness/AffinityAwarenessConnection.spec.js]: I think we should check `error.message`, not `error.stack`

File: spec/affinity_awareness/AffinityAwarenessConnection.spec.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                const badEndpoints = ['127.0.0.1:10900', '127.0.0.1:10901'];
+                const realEndpoints = TestingHelper.getEndpoints(SERVER_NUM);
+
+                for (const ep of realEndpoints)
+                    expect(badEndpoints).not.toContain(ep);
+
+                const client = TestingHelper.makeClient();
+                const cfg = new IgniteClientConfiguration(...badEndpoints).setConnectionOptions(false, null, true);
+
+                try {
```
Comment (ptupitsyn) [inline spec/affinity_awareness/AffinityAwarenessConnection.spec.js:61]: `expect.toThrow` is a cleaner way to assert exceptions

File: spec/affinity_awareness/AffinityAwarenessConnection.spec.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+const CACHE_NAME = '__test_cache';
+const SERVER_NUM = 3;
+
+describe('affinity awareness with checks of connection to cluster test suite >', () => {
+    beforeEach((done) => {
+        Promise.resolve().
+            then(async () => {
+                await TestingHelper.initClusterOnly(SERVER_NUM, true);
+            }).
+            then(done).
```
Comment (ptupitsyn) [inline spec/affinity_awareness/AffinityAwarenessConnection.spec.js:36]: Dots should be at the beginning of the line, not at the end:
`.then(done)`

File: spec/affinity_awareness/AffinityAwarenessMultipleConnections.spec.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    it('put with affinity awareness and unknown cache', (done) => {
+        Promise.resolve().
+            then(async () => {
+                const cache = await getCache(ObjectType.PRIMITIVE_TYPE.INTEGER, ObjectType.PRIMITIVE_TYPE.INTEGER, '__unknown_cache_359f72tg');
+                let key = 42;
+                try {
+                    await cache.put(key, key);
+                }
+                catch (error) {
+                    expect(error.toString()).toContain('Cache does not exist');
```
Comment (ptupitsyn) [inline spec/affinity_awareness/AffinityAwarenessMultipleConnections.spec.js]: Same as above - let's make sure `error.message` is good

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (isapego) [reply .gitignore:6]: Yeah, I know, but it was not put there by original author that knows more about node.js, so I thought that maybe there is a reason for that. So I decided to keep it that way for now.

File: spec/affinity_awareness/AffinityAwarenessConnection.spec.js (hunk 5 above)
Comment (isapego) [reply spec/affinity_awareness/AffinityAwarenessConnection.spec.js:36]: Well, it's the code style of original author. I don't think I should change it now.

File: spec/affinity_awareness/AffinityAwarenessConnection.spec.js (hunk 4 above)
Comment (isapego) [reply spec/affinity_awareness/AffinityAwarenessConnection.spec.js:61]: It is, but it does not catch anything by some reason.

File: spec/affinity_awareness/AffinityAwarenessConnection.spec.js (hunk 2 above)
Comment (isapego) [reply spec/affinity_awareness/AffinityAwarenessConnection.spec.js:50]: Again, this is how it is in the whole library. Not sure I should have done re-factoring in scope of this ticket.

File: spec/affinity_awareness/AffinityAwarenessConnection.spec.js (hunk 3 above)
Comment (isapego) [reply spec/affinity_awareness/AffinityAwarenessConnection.spec.js]: Good catch

File: spec/affinity_awareness/AffinityAwarenessMultipleConnections.spec.js (hunk 6 above)
Comment (isapego) [reply spec/affinity_awareness/AffinityAwarenessMultipleConnections.spec.js]: Fixed

File: spec/affinity_awareness/AffinityAwarenessConnection.spec.js (hunk 5 above)
Comment (ptupitsyn) [reply spec/affinity_awareness/AffinityAwarenessConnection.spec.js:36]: Oh wow, okay then.

File: spec/affinity_awareness/AffinityAwarenessConnection.spec.js (hunk 2 above)
Comment (ptupitsyn) [reply spec/affinity_awareness/AffinityAwarenessConnection.spec.js:50]: Ok

**Review by ptupitsyn — APPROVED** [review]

File: spec/affinity_awareness/AffinityAwarenessTestUtils.js [hunk 7]
```diff
@@ -58,6 +58,24 @@ class AffinityAwarenessTestUtils {
         throw 'All requests go to the same server when random was expected';
     }
 
+    static async testSameNode(cache) {
+        let key = 1337;
```
Comment (ptupitsyn) [inline spec/affinity_awareness/AffinityAwarenessTestUtils.js:62]: 😎

**Review by ptupitsyn — APPROVED** [review]

**Pull request merged**